### DISCUSSION TITLE
:put but only if does not exist

### DISCUSSION THREAD

**Discussion by dcz-self** [cozodb/cozo#138] [Q&A]
Currently, the documentation states that :put will overwrite the original keys if those match.

That's not great if I want to keep data from overwriting, e.g. if a resource exists under that name, I don't want to clobber it.

How to do this atomically in cozo? I tried to mess with the % syntax to do %if exists %then :put % else nothing, but that immediately crashed inside cozo.

**Comment by zh217** [comment]
If you are running the latest version, there is another set of mutations `:insert`, `:update` and `:delete` which does what you expect from SQL. You can use imperative ops, but it is no longer necessary and quite a bit faster. Yes the docs need more work.

**Comment by zh217** [comment]
And if by "crash" you mean "panic", can you share the script that caused it? All panics are bugs.

**Reply by dcz-self** [reply]
I do mean panic, but I don't have the code any more. It was similar to https://github.com/cozodb/cozo/issues/131 . I can't reproduce it because I don't understand the imperative documentation, and I keep getting syntax errors.
